rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,189 @@
+    "hostname": "www.altostrat.com",
+    "ip": "198.51.100.68",
+    "port": 443,
+    "url": "www.altostrat.com/images/logo.png"
+  }
+}
+```
+
+Events can be batched and sent in a JSON object that also includes the
+`customer_id`:
+
+```json
+{
+"customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+"events": [...]
+}
+```
+
+UDM events can be sumbitted as either JSON or Proto3. We will submit using JSON
+as this fits in easier with Vectors data model. The newly developed schema
+support for Vector will be used to ensure the JSON is formed according to the
+[Chronicle requirements](https://cloud.google.com/chronicle/docs/unified-data-model/format-events-as-udm).
+
+Vector has no internal knowledge of SIEM events so it will be up to the user to
+ensure the event is well formed - either the source will receive the data in the
+correct structure, or it will be transformed into such before reaching the sink.
+
+#### Unstructured log data events
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable.
+
+Another field `remove_log_type` allows the field used in the template to be
+removed from the event.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
+
+## Rationale
+
+There is significant user demand for sending data into Google Chronicle.
+
+## Drawbacks
+
+Adding a new sink does place an additional maintenance burden on the team. In
+particular since Chronicle is a third party service, developing integration
+tests to ensure the sink continues to work correctly is burdensome.
+
+## Prior Art
+
+## Alternatives
+
+An option that was considered was to use the HTTP sink and and enhance it to
+allow GCP authentication. This would nearly work for UDM events, but
+unfortunately since the payload is a JSON object with a `customer_id` field
+and an array of `events`, the structure is slightly different to what is
+possible with the HTTP sink.
```

Comment:
What would the cost/benefit calculation be of making the `http` sink more flexible, to support this use-case, as opposed to adding a custom sink?

One downside would be the loss of schema validation, which seems significant enough to me to encourage introducing a separate sink, but I want to make sure others feel the same, and we can't get away with improving the `http` sink instead.

Reply:
I am hesitant to do an improved `http` sink, at least initially. There's potentially a number of edge cases that could make things tricky since making it work for Chronicle may cause breaking changes to the existing sink, so we would need to be super cautious.

Once this sink is working in isolation it would be worth revisiting this idea. Then we could apply any lessons learned here to apply it to `http`.

Reply:
👍 yeah, I tend to agree. We can implement it separately first, and then try to fold them together.

Note that the batch size cap is 1 MB **uncompressed**.